### PR TITLE
workflows: make cockroach-microbench-ci as required

### LINKED ISSUE #106661: perf: avoid small performance regressions
**Is your feature request related to a problem? Please describe.**

We currently rely on changepoint detection and manual inspection on https://roachperf.crdb.dev/ to catch performance regressions. We are usually only able to claw back large regressions, which are bisected to a commit in a labor-intensive process (which we've been automating a bit, but still).

But looking at many graphs (like kv95) we see that we are continually backsliding. It's easy to see why: allocations are expensive, Go makes it hard to avoid accidental allocations, there is no forcing function in CI, et cetera.

**Describe the solution you'd like**

I'd like some way to avoid this continuous backsliding. My expectation is that the best way to achieve this is by preventing at least some classes of regressions in CI. For example, by comparing op/s on a small set of representative benchmarks against the parent commit, or by controlling for allocations (though regressions can also be introduced by just adding code!)

**Describe alternatives you've considered**

**Additional context**

Jira issue: CRDB-29662

### PR DESCRIPTION
**Pull request by sambhav-jain-16**
This change aims to make `cockroach-microbench-ci` a required step.
There is an addition of a label `performance-regression-expected` that
can be used to skip the comparison in case a regression is expected

Epic: https://cockroachlabs.atlassian.net/browse/CRDB-40772
Fixes: https://github.com/cockroachdb/cockroach/issues/106661

### CODE REVIEW

**Comment by sambhav-jain-16** [comment]
Tested this with both with label and without label

**Comment by sambhav-jain-16** [comment]
> Is the runtime of this new job longer than the current longest required jobs (unit tests and lint)?

Yes, the current runtime is ~31 min, which is longer than lint ~26 min

**Comment by dt** [comment]
> the current runtime is ~31 min

Any chance we could narrow the set of tests in the required job a bit and/or split it across jobs such that we aren't regressing on end-to-end time for ci?

**Comment by msbutler** [comment]
> could narrow the set of tests in the required job a bit

If a patch changes the backupccl package, i don't see the need to run the microbenchmarks in the storage package. Before merging any sort of microbenchmark test as required in ci, can we limit the benchmarks run to the to packages that were changed in the commit? 


I'm quite worried this change would lead to very flaky CI.

**Comment by msbutler** [comment]
I'm also skeptical this would add more test coverage, given the additional monetary cost this would add. Like, we already run these benchmarks once a night, right? If the owning team gets a failure the morning after they merged a problematic patch, they should be able to quickly bisect the regression.

**Review by rickystewart — CHANGES_REQUESTED** [review]
I agree with concerns about this change. We can't be making CI any longer to complete. I am also concerned about introducing spurious CI failures.

**Comment by sambhav-jain-16** [comment]
> Any chance we could narrow the set of tests in the required job a bit and/or split it across jobs such that we aren't regressing on end-to-end time for ci?

The script uses a single `BenchmarkKV` from the sql package [ref],(https://github.com/cockroachdb/cockroach/blob/eac4efe9be039d149aa8539362a1fd241c0b06d2/build/github/cockroach-microbench-ci.sh#L49) there's no way to split it since it is a single microbenchmark.

> If a patch changes the backupccl package, i don't see the need to run the microbenchmarks in the storage package. Before merging any sort of microbenchmark test as required in ci, can we limit the benchmarks run to the to packages that were changed in the commit?
I'm quite worried this change would lead to very flaky CI.

Same as above, since `BenchmarkKV` covers end-to-end benchmark for the in-memory cockroach, so we can't limit per package.

As per the flakiness part, this has been addressed by analysing the variance of the subsequent runs which were plotted [here](https://lookerstudio.google.com/reporting/d707c4cc-5b86-490f-891f-b023766bb89f/page/n1tCE). There were also changes by @nvanbenschoten to reduce the variance https://github.com/cockroachdb/cockroach/pull/131711.

Apart from the above changes, I have analysed past 1 month of microbenchmark and @srosenberg and me arrived at a threshold value of `20%` for a regression, this change was merged [here](https://github.com/cockroachdb/cockroach/pull/134810)

> We can't be making CI any longer to complete.

This can be addressed by reducing the no of count of the benchmark run but that will result in more variance in the benchmark runs. @srosenberg wdyt about this? If that's the trade-off we want to take for faster ci.

> Like, we already run these benchmarks once a night, right? If the owning team gets a failure the morning after they merged a problematic patch, they should be able to quickly bisect the regression.

https://github.com/cockroachdb/cockroach/issues/106661 mentions how roachperf is not been able to detect small regressions that have been incrementally added by PRs, the ci job aims to address this by stopping the PR from getting merge itself.

**Comment by rickystewart** [comment]
> there's no way to split it since it is a single microbenchmark.

Obviously, since `BenchmarkKV` is nothing more than a loop with 4 iterations, you could extract it into 4 different benchmarks.

**Comment by rickystewart** [comment]
> Apart from the above changes, I have analysed past 1 month of microbenchmark and @srosenberg and me arrived at a threshold value of 20% for a regression, this change was merged https://github.com/cockroachdb/cockroach/pull/134810

How many spurious failures have occurred since this change was made? Have any legitimate failures been observed?

**Comment by dt** [comment]
> https://github.com/cockroachdb/cockroach/issues/106661 mentions how roachperf is not been able to detect small regressions that have been incrementally added by PRs, the ci job aims to address this by stopping the PR from getting merge itself.

I'm not quite following this? If small regressions are hard to notice due to being small, how does looking for them before vs after merge change the scale of the regression we can reliably detect? It seems like f we're able to mechanically determine if one SHA is a regression compared to another SHA with high confidence -- such that we can run it on the PR's SHA vs its mergebase -- then it seems like we, could also run this once or twice a day comparing to the prior day's HEAD SHA to file an issue (with the delta commit log) too? Does this really need to be in the PR path?

**Comment by rickystewart** [comment]
I also don't know if I would consider 20% to be a "small" regression. It also seems like a regression of that magnitude would be clearly visible in the nightlies.

**Comment by sambhav-jain-16** [comment]
> > #106661 mentions how roachperf is not been able to detect small regressions that have been incrementally added by PRs, the ci job aims to address this by stopping the PR from getting merge itself.
> 
> I'm not quite following this? If small regressions are hard to notice due to being small, how does looking for them before vs after merge change the scale of the regression we can reliably detect? It seems like if we're able to mechanically determine if one SHA is a regression compared to another SHA with high confidence -- such that we can run it on the PR's SHA vs its mergebase -- then it seems like we could also just do so once or twice a day, comparing the current HEAD SHA to last one and file an issue (with the delta commit log) if it has regressed?
> 
> Does the comparison really need to be done for every PR?

The idea of adding doing comparison on every PR is that it forces the engineer to fix regressions in the PR itself rather than again waiting for the nightlies to run and then again go ahead analyse the code to detect the issue and fix. This make the turn around time higher that making the engineer fix the issues in the current PR itself.

**Comment by dt** [comment]
> forces the engineer to fix regressions in the PR itself

Do we need to _force_ engineers to do this? Is there evidence that if we present engineers with _reliable_ information that they just merged a significant regression that they won't act quickly to fix/revert/etc?

> waiting for the nightlies to run ... turn around time

While that's true, it isn't clear that that is _too much_ turn around time, or that cutting that down is worth the cost (either in $ or more importantly in productivity due to a slower CI pipeline). That said, if we really wanted something quicker I think we still have options that aren't to be in the blocking path of every PR, e.g. I believe we already have a CI that runs every time master's SHA has changed ("merge to master" or "publish bleeding edge" or something? dev-inf/rel-eng would know the details) in which the list of candidate changes compared to last run would be relatively small (just what was in the bors batch).

**Comment by srosenberg** [comment]
> Do we need to force engineers to do this? Is there evidence that if we present engineers with reliable information that they just merged a significant regression that they won't act quickly to fix/revert/etc?

I guess I would rephrase the above differently. The operation assumption is that the specifically chosen microbenchmarks are fairly _stable_, and the asserted performance difference denotes a sizable regression, with high probability. If that's the case, then how is this different from forcing engineers that the unit tests pass? Catching a regression after an offending PR is merged, translates to doing more (infra.) work. When would you not want to block PR, after learning that it contains a sizable performance regression?

**Comment by rickystewart** [comment]
> The operating assumption is that the specifically chosen microbenchmarks are fairly stable, and the asserted performance difference denotes a sizable regression, with high probability. If that's the case, then how is this different from forcing engineers that the unit tests pass?

One difference might be that unit tests fail regularly and are frequently broken by routine PR's to `cockroach`. If we didn't check unit tests in CI, then a large proportion of non-trivial PR's would result in some failure that would have to be resolved after the fact by the engineer. So running unit tests in CI is a huge boon for developer productivity in that way.

How often do single-PR performance regressions to the tune of 20% in a single microbenchmark occur? Once per month, week, year? Setting aside developer experience concerns (e.g. are there false positives, etc.), if this error case does not frequently occur, it is very unlikely that this is even worth the added CI cost.

**Comment by srosenberg** [comment]
> One difference might be that unit tests fail regularly and are frequently broken by routine PR's to `cockroach`. If we didn't check unit tests in CI, then a large proportion of non-trivial PR's would result in some post-merge failure that would have to be resolved after the fact by the engineer. So running unit tests in CI is a huge boon for developer productivity in that way.
> 

The same argument could be made for microbenchmarks, which can be seen as unit tests that assert on performance. We know from experience, that some performance bugs can quickly turn into unavailability bugs. Fundamentally, I don't see a difference.

> How often do single-PR performance regressions to the tune of 20% in a single microbenchmark occur? Once per month, week, year? Setting aside developer experience concerns (e.g. are there false positives, etc.), if this error case does not frequently occur, it is very unlikely that this is even worth the added CI cost.

That's a reasonable question. I suspect it's the primary reason why few places succeed in enforcing this check on PRs; that and a lack of discipline to address both correctness and performance in the same PR.

Looking at similar microbenchmarks in [Go](https://github.com/golang/go/) (e.g., `CockroachDBkv95/nodes=1-8`), over last 7 days, I spot several commits with regressions larger than 20%. We can see what Sambhav digs up from our own runs.

[1] https://perf.golang.org/dashboard/?benchmark=regressions&unit=&repository=go&branch=master&platform=all&days=7&end=2024-11-26T03%3A36

**Comment by sambhav-jain-16** [comment]
> I will get the data tomorrow IST and post it here.

I ran a script from 9th oct to 9th nov. 
This is the output of the comparison
[output_20.txt](https://github.com/user-attachments/files/17913396/output_20.txt)

These are some of the potential PRs that were caught by the job:

https://github.com/cockroachdb/cockroach/pull/134325
https://github.com/cockroachdb/cockroach/pull/134183
https://github.com/cockroachdb/cockroach/pull/134019
https://github.com/cockroachdb/cockroach/pull/133686
https://github.com/cockroachdb/cockroach/pull/133298
https://github.com/cockroachdb/cockroach/pull/132603
https://github.com/cockroachdb/cockroach/pull/132766
https://github.com/cockroachdb/cockroach/pull/131828

**Comment by tbg** [comment]
We're about to go through the codebase and fix many minor regressions as part of our perf efforts. Not even forcing 20% regressions to be reckoned with in the moment is incompatible with that effort. I'm a strong +1 to adding something in the critical path of the development workflow that at least requires a manual override before a merge.

Reading between the lines in this thread I get a sense that folks are worried that this is just going to be noisy and unhelpful. As is, that might be true. But I believe that done right, this check does not have to be "yet another annoying thing" in your day-to-day work. Instead, it could be a valuable tool that we love to have around, one that's helpful and high-signal and one that automatically does for us twelve steps that we'd otherwise have to trudge through the next day.

I've been running microbenchmarks _frequently_ recently and I don't think this needs to take 31 minutes, unless there is some insurmountable problem where getting the build done already takes 20. I'm happy to work with test-eng to look into the benchmark selection and ways to squeeze out more signal.

My vote would be making the check faster and the ergonomics better first, and then making it mandatory. For example, I spot a few PRs in [this list](https://github.com/cockroachdb/cockroach/pull/135899#issuecomment-2499722307) that I'm pretty sure had exactly _no_ impact on performance. But I definitely am not going to check because that would be quite annoying to actually carry out - an ergonomics issue. It needs to be so pleasant that you'll _want_ to take a look.

----

> I'm also skeptical this would add more test coverage. Like, we already run these benchmarks once a night, right? If the owning team gets a failure the morning after they merged a problematic patch, they should be able to quickly bisect the regression.

It's quite expensive to go back after the fact and revisit code that was merged before and run benchmarks. Saying that we just "don't care" about perf regressions during CI also communicates, well, that it's just not that important. I'd posit that regressions are often small fixes in before the merge - or at least can be accepted consciously - but after the merge they amount to an annoying and expensive context switch that in practice we don't very proactively do.

----

I agree with previous comments that the regression check must not be permitted to increase overall CI length. I would simply call this a requirement.

----

If I understand correctly, currently (after digging through a stack trace) this is how the job fails if it catches a regression:
```
I241126 05:29:15.389634 1 (gostd) main.go:252  [-] 1 +  | Metric: sec/op, Benchmark: pkg/sql→KV/Delete/SQL/rows=100, Change: +21.82%
```

I'd really want the `benchdiff` printout, CPU and mem profiles ready to be investigated, and commands I can just copy-paste. This should all be doable with some elbow grease and I'm happy to guide.

By the way, I see that the current check makes a new run only for the commit to be tested, and pulls the old numbers off S3. I think that is a mistake - this introduces correlated bias into the results that (I think) significantly reduces the chances of keeping noise low. Instead, we should stash the binaries on S3 and run both the old and new code on the same machine for a direct comparison, at least for CPU time measurements. (Even reusing binaries can be problematic since one binary may be more efficient than another one simply due to some random compiler choices, but I think that we will have to live with).
I think we also want to look into more stable proxies for performance, for example allocation counts.

**Comment by sambhav-jain-16** [comment]
I added a change to run that removes the bound of running the  microbenchmark on one cpu. This has decreased the ci job from 31 min to ~22 min.

>  Instead, we should stash the binaries on S3 and run both the old and new code on the same machine for a direct comparison, at least for CPU time measurements.

We considered this while implementing the first time but this can further increase the CI job time, since I can't really run both benchmarks at the same time on a same machine. Therefore, this solution was designed with the thought that the CI jobs run on same machine types. I do agree that this still can have variance (as I have seen while I ran a single benchmark multiple times of a same machine without any changes), this was done to save CI time. 

> I'd really want the benchdiff printout, CPU and mem profiles ready to be investigated, and commands I can just copy-paste. This should all be doable with some elbow grease and I'm happy to guide.

I think this can be done, We can sync up to see how this can be implemented, I'm happy to work on this if the engineers feel this will be helpful to debug.

**Comment by dt** [comment]
> strong +1 to adding something in the critical path of the development workflow

The thing that strikes me as a leap is to go straight to deeming that it must be in the *critical* path when we haven't tried first tried to see if our problem is solved by it being on the development workflow path _at all_, such as in the form a reliable post-merge notification of a detected regression.

>  If that's the case, then how is this different from forcing engineers that the unit tests pass?

Because a broken unit test harms the productivity of _other engineers_: if the code in some library doesn't compile or does not produce the correct results when called, I cannot build and test the behavior of features in *my* library and parts of the codebase if those parts can't compile their dependencies or are getting incorrect inputs from them. Breaking a core library could bring all development across all teams to a halt. Whereas if a library becomes x% slower for one night, that is far less likely to have a broad impact on productivity across the organization.

My impression is that we believe that we've had quiet unnoticed perf regressions sneaking in but I believe we also all agree that this is because they simply were not identified at the time / brought to the attention of the author of the change that introduced them.

I think getting the time of this job per run down is essential no matter when we run it. If we want to run it per-merge, it also must be comparing to the merge-base, not current HEAD -- otherwise every time someone merges an optimization, all other wholly unrelated PRs will start to flake as regressions. (Alternatively it could benchmark the new SHA as rebased on head but none of our CIs do that and it would likely be confusing to engineers to have one odd job doing that).

But even if you could get the job down to, say, 20 minutes, it will still make the CI pipeline flakier just because _every_ job, but especially such a long running one, can be preempted these days. So the more jobs we elect to run on the critical path increases the higher the chance of a flake on that path, _even if the jobs themselves are 100% reliable_, so we have to be sure that the benefit of doing so justifies the cost. In this case, it really seems like we can get the lion's share of that benefit from a post-merge check can't we?

**Comment by msbutler** [comment]
I've been convinced that a pre-merge microbenchmark job would be good to add, so long as it doesn't effect general productivity. 

One reason I came out swinging against this change is because my team doesn't really rely on microbenchmarks, so as Tobi has said, I really fear this change could bog down time to merge. Furthermore,  I think we _already_ run some sort of bench job pre merge in extended ci it is optional. This job seems incredibly flaky and is often the bottleneck to extended CI runs. Since many engineers wait for extended CI to complete before merging, we already have a bench job slowing down time to merge! @tbg Are kv engineers using this extended ci bench job in their workflow? 

Since @sambhav-jain-16 's work seems much more targeted, I'm hoping one outcome of it will be to sunset extended ci's bench job.

**Comment by tbg** [comment]
> The thing that strikes me as a leap is to go straight to deeming that it must be in the critical path when we haven't tried first tried to see if our problem is solved by it being on the development workflow path at all, such as in the form a reliable post-merge notification of a detected regression.

Let's table the discussion on whether this should be required. Personally, I don't want to require it, at least not in the "you can't merge until you fixed it" kind of sense. But I do think we should run these in the critical path and make the result visible, i.e. you'd have to ack your way through a regression to merge it. (Plus, you get the nice feeling of having done a good job when you actually made things faster). The important part is to _do it well_ so that it's a pleasure to engage with. I don't want to make this the selling point here, but I think putting the perf tooling into people's PRs will also be a real boost to the perf culture at CRL.

I'd argue that the async path is what we have tried so far. For example, [here](https://cockroachlabs.slack.com/archives/C05SHB3GAGP/p1732527632687529), every week. But it is fundamentally a slow path. It's like fixing a bug after the merge, except - as you point out - the bug doesn't "really hurt that much" and surely as an author you have other things to do, which in my estimate puts the chances of smaller issues being fixed at approximately zero percent.

We can't run every benchmark in CI and expect that to be useful (looking at you, bench job), so we will need an "async path" for most benchmarks anyway. I am all for that, but that is not this.

>  Furthermore, I think we already run some sort of bench job pre merge in extended ci (which is optional). This job seems incredibly flaky and is often the bottleneck to extended CI runs. Since many engineers wait for extended CI to complete before merging, we already have a bench job slowing down time to merge! @tbg Are kv engineers using this extended ci bench job in their workflow?

I've never seen anyone use this job. I'm also not sure what it's there for: certainly it's not providing any value to the CI pipeline; you can't even see the results and benchmarks are meaningless without a reference point. It maybe convinces us that each of the benchmark actually compiles and can run without failing? We should just do that in nightlies, since it's rare to break one and we compile them already when running the regular tests. If they fail in nightly, they should just file issues.

As a CI job, I'd remove this ASAP.

**Comment by dt** [comment]
> I'd argue that the async path is what we have tried so far. For example, [slack bot that does not identify specific changes or notify their authors], every week. But it is fundamentally a slow path.

I think this is conflating the effect of two variables and has some element of a correlation-vs-causation blurring: our path so far has indeed been async but has also been extremely non-specific. If I merge a regression, it is up to me see a message from a slackbot that doesn't mention me or my change some days later, and then think to go check if that might have been me. That fact that that is after I merge versus before I merge I don't think is the primary reason it isn't acted on, but rather the fact it is so non-specific and thus unactionable. 

I'd argue that we haven't yet tried a path where we give the authors of regressions specific, actionable feedback about their changes, either before _or after_ they merge those changes, so it is isn't clear to me that we can conclude we have tried "the" async path and can conclude it is a dead end based on our lack of success so far?

But all that said, if this new CI job can be made <10mins, is strictly comparing to the merge base, and doesn't flake -- and we're ready to disable it as soon as any of those are observed not to hold -- then isn't a hill I'm prepared to die on.

**Comment by rickystewart** [comment]
> I added a change to script that removes the bound of running the microbenchmark on one cpu.

Doesn't this jeopardize the benchmark results in some way? Generally people choose to run the benchmarks on a constant number of CPU's with limited parallelism to make the benchmark results more accurate. Did you do some sort of validation that the benchmark you're testing has not been affected by this change?

**Comment by rickystewart** [comment]
If you're going to change some fundamental aspect of how the benchmarks are run to make the runtime shorter, I would prefer to make that change in a separate PR, then you can watch it for ~a week to determine that we haven't made the job flakier before making it required in CI.

**Comment by rickystewart** [comment]
> These are some of the potential PRs that were could have been caught by the job:

How did you produce the list? I see the first PR is mine, #134325. But this PR did not introduce any major performance regression (that I have been made aware of).

**Comment by sambhav-jain-16** [comment]
> Did you do some sort of validation that the benchmark you're testing has not been affected by this change?

Yes, this was done in the outputs I posted above, they were ran without fixating the cpu and showed the similar characteristics. I added that change in this PR itself to test it out since looks like anyways I need to make some fundamental changes in the PR as some of the folks mentioned in the PR.
Also, just to point out again, anything that has been done in this PR before also has been tested and this job has ran for some time now to collect data, and the numbers that has been arrived are after analysing that data which I posted [here](https://github.com/cockroachdb/cockroach/pull/135899#issuecomment-2498743508). These numbers have resulted in false positives and people seem to be concerned about the flakiness, which I intend to address going forward.

> If you're going to change some fundamental aspect of how the benchmarks are run to make the runtime shorter, I would prefer to make that change in a separate PR, then you can watch it for ~a week to determine that we haven't made the job flakier before making it required in CI.

There are suggestions by @tbg on using different benchmark altogether that toby has asked me to explore. I'm also going to wait for @srosenberg to comment on what he thinks about the suggestions pointed out above.

**Comment by sambhav-jain-16** [comment]
> How did you produce the list? I see the first PR is mine, https://github.com/cockroachdb/cockroach/pull/134325. But this PR did not introduce any major performance regression (that I have been made aware of).

Were microbench ran against the change to see if there's a regression? Again, this change doesn't intend to capture a "major" regression, but smaller regressions that might not get caught in the nightlies.

List was accumulated from the bucket that stores the output of ci job itself and the same `roachprod-microbench compare` command was ran on subsequent runs.
The compare command list the commits in which there was a regression from the previous master commit

**Comment by tbg** [comment]
> I'd argue that we haven't yet tried a path where we give the authors of regressions specific, actionable feedback about their changes, either before or after they merge those changes, so it is isn't clear to me that we can conclude we have tried "the" async path and can conclude it is a dead end based on our lack of success so far?

I can't really disagree with you. I'd like to see something useful, if it's right after the PR in a slack message - fine by me too. Useful comes first. I don't think what we have right now is useful yet.

> Doesn't this jeopardize the benchmark results in some way? Generally people choose to run the benchmarks on a constant number of CPU's with limited parallelism to make the benchmark results more accurate. Did you do some sort of validation that the benchmark you're testing has not been affected by this change?

Perhaps surprisingly, giving Go just one CPU is actually a bad idea as far as I understand. These benchmarks run entire servers. Servers have async work, which, at one CPU, interrupts foreground work. So you want to have at least a few processors so that these benchmarks - which still only do sequential client requests - end up with predictable variance. Basically they measure the speed at which a single client can go against a system that has enough slack. This is useful. It's different from measuring how fast the system can go when maximally loaded, which is also useful but not what we're after here, as it's typically less stable and much more sensitive to the environment. Either way, running with one CPU is neither. It's "what if we ran all of CRDB on one CPU". We know that isn't great, and we don't care about it either.

**Comment by tbg** [comment]
Re [this](https://github.com/cockroachdb/cockroach/pull/135899#issuecomment-2501273878), I also have my doubts about this list. For example https://github.com/cockroachdb/cockroach/pull/133298 certainly did not introduce any regression; there aren't even any production code changes in there unless requests are being traced. So I think this shows us that we need to do better than what we have as of this PR. We're really nowhere close to deciding where these checks should live - let's make them useful first and then work up a stronger conviction of what their fate should be.

Here's `benchdiff ./pkg/sql/tests -b -r 'Sysbench/SQL/3node/oltp_read_write' -d 1000x -c 5` for the above PR which shows very similar performance.

<details><summary>Benchmarking output</summary>
<p>

```
benchdiff ./pkg/sql/tests -b -r 'Sysbench/SQL/3node/oltp_read_write' -d 1000x -c 5
checking out '7342c63'
building benchmark binaries for '7342c63' [bazel=true] 1/1 /
checking out 'e77ae9b'
building benchmark binaries for 'e77ae9b' [bazel=true] 1/1 |

  pkg=1/1 iter=5/5 cockroachdb/cockroach/pkg/sql/tests /

name                                   old time/op    new time/op    delta
Sysbench/SQL/3node/oltp_read_write-24    14.5ms ± 3%    14.5ms ± 2%   ~     (p=0.690 n=5+5)

name                                   old alloc/op   new alloc/op   delta
Sysbench/SQL/3node/oltp_read_write-24    2.56MB ± 1%    2.57MB ± 2%   ~     (p=1.000 n=5+5)

name                                   old allocs/op  new allocs/op  delta
Sysbench/SQL/3node/oltp_read_write-24     17.4k ± 1%     17.4k ± 1%   ~     (p=1.000 n=5+5)
```

</p>
</details>

**Comment by rickystewart** [comment]
> > How did you produce the list? I see the first PR is mine, https://github.com/cockroachdb/cockroach/pull/134325. But this PR did not introduce any major performance regression (that I have been made aware of).

> Were microbenchmarks run against the change to see if there's a regression?

You're the one claiming this PR introduced a performance regression, so where is the data in support of this?

**Comment by rickystewart** [comment]
> Perhaps surprisingly, giving Go just one CPU is actually a bad idea as far as I understand.

This is a tangent, but `dev bench` defaults to one CPU these days (as of `a6a27cd7d38dcab3a08bd5de1c2d7766c26c46b7`) as people were surprised/upset it didn't do that for you automatically. So if there is some clear best practice that says the opposite, I would recommend socializing that idea, as I don't know if there is broad agreement/knowledge that that's the case.

**Comment by sambhav-jain-16** [comment]
> You're the one claiming this PR introduced a performance regression, so where is the data in support of this?

This is the output of merge of #134325 [(link from the bucket)](cockroach-microbench-ci/master/8c34a77f5c6e806b359f20a6fd843817b1cc9d5e.log)
[file](https://github.com/user-attachments/files/17923880/8c34a77f5c6e806b359f20a6fd843817b1cc9d5e.log)

This is the previous commit on master before it [link from the bucket](cockroach-microbench-ci/master/d0ff9c702fbf391c4a2e080b81a6c64400a39bad.log)
[file](https://github.com/user-attachments/files/17923887/8c34a77f5c6e806b359f20a6fd843817b1cc9d5e.log)

and this is the output of comaprison

```
❯ roachprod-microbench compare 1 0 --threshold 20
I241126 17:44:15.712839 1 (gostd) main.go:252  [-] 1  ERROR: there are benchmark regressions of > 20.00% in the following packages:
I241126 17:44:15.712839 1 (gostd) main.go:252  [-] 1 +(1) attached stack trace
I241126 17:44:15.712839 1 (gostd) main.go:252  [-] 1 +  -- stack trace:
I241126 17:44:15.712839 1 (gostd) main.go:252  [-] 1 +  | main.(*compare).compareUsingThreshold
I241126 17:44:15.712839 1 (gostd) main.go:252  [-] 1 +  | 	main/pkg/cmd/roachprod-microbench/compare.go:351
I241126 17:44:15.712839 1 (gostd) main.go:252  [-] 1 +  | main.makeCompareCommand.func1
I241126 17:44:15.712839 1 (gostd) main.go:252  [-] 1 +  | 	main/pkg/cmd/roachprod-microbench/main.go:188
I241126 17:44:15.712839 1 (gostd) main.go:252  [-] 1 +  | github.com/spf13/cobra.(*Command).execute
I241126 17:44:15.712839 1 (gostd) main.go:252  [-] 1 +  | 	github.com/spf13/cobra/external/com_github_spf13_cobra/command.go:916
I241126 17:44:15.712839 1 (gostd) main.go:252  [-] 1 +  | github.com/spf13/cobra.(*Command).ExecuteC
I241126 17:44:15.712839 1 (gostd) main.go:252  [-] 1 +  | 	github.com/spf13/cobra/external/com_github_spf13_cobra/command.go:1044
I241126 17:44:15.712839 1 (gostd) main.go:252  [-] 1 +  | github.com/spf13/cobra.(*Command).Execute
I241126 17:44:15.712839 1 (gostd) main.go:252  [-] 1 +  | 	github.com/spf13/cobra/external/com_github_spf13_cobra/command.go:968
I241126 17:44:15.712839 1 (gostd) main.go:252  [-] 1 +  | main.main
I241126 17:44:15.712839 1 (gostd) main.go:252  [-] 1 +  | 	main/pkg/cmd/roachprod-microbench/main.go:251
I241126 17:44:15.712839 1 (gostd) main.go:252  [-] 1 +  | runtime.main
I241126 17:44:15.712839 1 (gostd) main.go:252  [-] 1 +  | 	GOROOT/src/runtime/proc.go:271
I241126 17:44:15.712839 1 (gostd) main.go:252  [-] 1 +  | runtime.goexit
I241126 17:44:15.712839 1 (gostd) main.go:252  [-] 1 +  | 	src/runtime/asm_arm64.s:1222
I241126 17:44:15.712839 1 (gostd) main.go:252  [-] 1 +Wraps: (2) there are benchmark regressions of > 20.00% in the following packages:
I241126 17:44:15.712839 1 (gostd) main.go:252  [-] 1 +  |
I241126 17:44:15.712839 1 (gostd) main.go:252  [-] 1 +  | Package: pkg/sql
I241126 17:44:15.712839 1 (gostd) main.go:252  [-] 1 +  | Metric: sec/op, Benchmark: pkg/sql→KV/Delete/SQL/rows=100, Change: +21.82%
I241126 17:44:15.712839 1 (gostd) main.go:252  [-] 1 +Error types: (1) *withstack.withStack (2) *errutil.leafError
```

I hope this answers your question

**Comment by rickystewart** [comment]
OK, I ran the benchmark in question at `d0ff9c702fbf391c4a2e080b81a6c64400a39bad` and `8c34a77f5c6e806b359f20a6fd843817b1cc9d5e` ("after"). I could not reproduce any 20% regression introduced in `8c34a77...`. Can you reproduce this or does it indeed seem like this check is flaky?

```
before

BenchmarkKV/Delete/SQL/rows=100       	    1107	    975721 ns/op	  479261 B/op	    4988 allocs/op
BenchmarkKV/Delete/SQL/rows=1000      	     100	  16787909 ns/op	10634184 B/op	   56985 allocs/op
BenchmarkKV/Delete/SQL/rows=10000     	       7	 177139738 ns/op	142393733 B/op	  693768 allocs/op

BenchmarkKV/Delete/SQL/rows=100       	    1171	   1009993 ns/op	  486320 B/op	    4978 allocs/op
BenchmarkKV/Delete/SQL/rows=1000      	     100	  16331729 ns/op	10586844 B/op	   56972 allocs/op
BenchmarkKV/Delete/SQL/rows=10000     	       8	 172847656 ns/op	141608931 B/op	  668106 allocs/op

BenchmarkKV/Delete/SQL/rows=100       	     944	   1231771 ns/op	  488684 B/op	    5008 allocs/op
BenchmarkKV/Delete/SQL/rows=1000      	     100	  17260990 ns/op	10626778 B/op	   57001 allocs/op
BenchmarkKV/Delete/SQL/rows=10000     	       7	 178148024 ns/op	141316301 B/op	  680248 allocs/op

after
BenchmarkKV/Delete/SQL/rows=100       	    1096	    989900 ns/op	  480403 B/op	    4986 allocs/op
BenchmarkKV/Delete/SQL/rows=1000      	     100	  16811722 ns/op	10669168 B/op	   58635 allocs/op
BenchmarkKV/Delete/SQL/rows=10000     	       8	 186308281 ns/op	148710421 B/op	  681618 allocs/op

BenchmarkKV/Delete/SQL/rows=100       	    1086	   1036353 ns/op	  481556 B/op	    4987 allocs/op
BenchmarkKV/Delete/SQL/rows=1000      	     100	  16586513 ns/op	10619645 B/op	   57931 allocs/op
BenchmarkKV/Delete/SQL/rows=10000     	       7	 180607661 ns/op	143315194 B/op	  695340 allocs/op

BenchmarkKV/Delete/SQL/rows=100       	    1038	    983737 ns/op	  487190 B/op	    4993 allocs/op
BenchmarkKV/Delete/SQL/rows=1000      	     100	  16540953 ns/op	10630585 B/op	   57038 allocs/op
BenchmarkKV/Delete/SQL/rows=10000     	       8	 181020609 ns/op	146163115 B/op	  670403 allocs/op
```

**Comment by rickystewart** [comment]
Asking my above questions in a different way, can someone point to *one* PR that 1) introduced a legitimate, *reproducible* benchmark regression 2) in `BenchmarkKV` 3) that is more than 20%+? When is the last time this occurred? Is it within the last year?

It is difficult to square the idea that this is a valuable check that will protect us from dangerous changes when the check is clearly flaky and producing false positives, and when we have difficulty producing even a single example of the kind of regression that we are trying to protect against.

Of course, what is being done is challenging -- it's hard to validate with certainty that performance has not regressed. So I am not trying to suggest that anything that's been done, has been done wrong, or could necessarily have been done better. However, if we are going to spend time and $$$ performing checks in CI, we need to have a reasonable level of confidence that 1) the check will not produce false positives and will have predictable, reproducible output and 2) the check will have a reasonable chance of failing for any given PR, i.e., there is some value/signal in the check.  I don't believe either of these two criteria have been met in this case.

**Comment by tbg** [comment]
@ricky the check is almost certainly flaky. This is not surprising, since it compares artifacts from different machines. Another "incorrect" regression is [here](https://github.com/cockroachdb/cockroach/pull/135899#issuecomment-2501428918).

I think it's clear that we can't merge this PR as is. I would recommend closing it and continuing the discussion on https://github.com/cockroachdb/cockroach/issues/106661, which can result in a better PR whose goal likely won't be to block merges, but to provide useful signal on performance changes to PR authors as part of PR CI. I made a concrete suggestion [here](https://github.com/cockroachdb/cockroach/issues/106661#issuecomment-2513991169).

**Comment by sambhav-jain-16** [comment]
Closing this PR in accordance to the above discussion.
Further discussions will be on https://github.com/cockroachdb/cockroach/issues/106661

**Pull request closed without merging**
